content/knowledge-base/guides/write-a-composition-function-in-go.md

```diff
@@ -0,0 +1,826 @@
+									"forProvider": {
+										"region": "us-east-2"
+									}
+								}
+							}`)},
+							"xbuckets-test-bucket-b": {Resource: resource.MustStructJSON(`{
+								"apiVersion": "s3.aws.upbound.io/v1beta1",
+								"kind": "Bucket",
+								"metadata": {
+									"annotations": {
+										"crossplane.io/external-name": "test-bucket-b"
+									}
+								},
+								"spec": {
+									"forProvider": {
+										"region": "us-east-2"
+									}
+								}
+							}`)},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	for name, tc := range cases {
+		t.Run(name, func(t *testing.T) {
+			f := &Function{log: logging.NewNopLogger()}
+			rsp, err := f.RunFunction(tc.args.ctx, tc.args.req)
+
+			if diff := cmp.Diff(tc.want.rsp, rsp, protocmp.Transform()); diff != "" {
+				t.Errorf("%s\nf.RunFunction(...): -want rsp, +got rsp:\n%s", tc.reason, diff)
+			}
+
+			if diff := cmp.Diff(tc.want.err, err, cmpopts.EquateErrors()); diff != "" {
+				t.Errorf("%s\nf.RunFunction(...): -want err, +got err:\n%s", tc.reason, diff)
+			}
+		})
+	}
+}
+```
+{{</expand>}}
+
+Run the unit tests using the `go test` command:
+
+```shell
+go test -v -cover .
+=== RUN   TestRunFunction
+=== RUN   TestRunFunction/AddTwoBuckets
+--- PASS: TestRunFunction (0.00s)
+    --- PASS: TestRunFunction/AddTwoBuckets (0.00s)
+PASS
+coverage: 52.6% of statements
+ok      github.com/negz/function-xbuckets       0.016s  coverage: 52.6% of statements
+```
+
+You can preview the output of a Composition that uses this function using
+the Crossplane CLI. You don't need a Crossplane control plane to do this.
+
+Create a directory under `function-xbuckets` named `examples`, and add the
+three files `xr.yaml`, `composition.yaml`, and `functions.yaml`.
+
+{{<expand "The xr.yaml, composition.yaml and function.yaml files">}}
+
+You can recreate the output below using by running `crossplane beta render` with
+these files.
+
+The `xr.yaml` file contains the composite resource to render:
+
+```yaml
+apiVersion: example.crossplane.io/v1
+kind: XBuckets
+metadata:
+  name: example-buckets
+spec:
+  region: us-east-2
+  names:
+  - crossplane-functions-example-a
+  - crossplane-functions-example-b
+  - crossplane-functions-example-c
+```
+
+The `composition.yaml` file contains the Composition to use to render the
+composite resource:
+
+```yaml
+apiVersion: apiextensions.crossplane.io/v1
+kind: Composition
+metadata:
+  name: create-buckets
+spec:
+  compositeTypeRef:
+    apiVersion: example.crossplane.io/v1
+    kind: XBuckets
+  mode: Pipeline
+  pipeline:
+  - step: create-buckets
+    functionRef:
+      name: function-xbuckets
+```
+
+The `functions.yaml` file contains the Functions the Composition references in
+its pipeline steps:
+
+```yaml
+apiVersion: pkg.crossplane.io/v1beta1
+kind: Function
+metadata:
+  name: function-xbuckets
+  annotations:
+    render.crossplane.io/runtime: Development
+spec:
+  # The CLI ignores this package when using the Development runtime.
```

Comment:
ah, i had these annotations in a functions.yaml file i was using for recent testing - are they no longer needed?
```
  annotations:
    # render will try to talk to your Function at localhost:9443
    render.crossplane.io/runtime: Development
    render.crossplane.io/runtime-development-target: localhost:9443
```